pkg/scheduler/scheduler_test.go

```diff
@@ -61,10 +72,88 @@ func TestForget(t *testing.T) {
 				})
 				queue.Add(test.obj, test.duration)
 				queue.Forget(test.obj)
-				time.Sleep(test.duration * 2)
+				after.warp(test.duration * 2)
 			}
 		}(test))
 	}
 
 	wg.Wait()
 }
+
+// TestConcurrentAdd checks that if we add the same item concurrently, it
+// doesn't end up hitting a data-race / leaking a timer.
+func TestConcurrentAdd(t *testing.T) {
+	after := newMockAfter()
+	afterFunc = after.AfterFunc
+	var wg sync.WaitGroup
+	queue := NewScheduledWorkQueue(func(obj interface{}) {
+		t.Fatalf("should be called, but was called with %v", obj)
+	})
+
+	for i := 0; i < 1000; i++ {
+		wg.Add(1)
+		go func() {
+			queue.Add(1, 1*time.Second)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	queue.Forget(1)
+	after.warp(5 * time.Second)
+}
+
+type timerQueueItem struct {
+	f       func()
+	t       time.Time
+	run     bool
+	stopped bool
+}
+
+func (tq *timerQueueItem) Stop() bool {
+	stopped := tq.stopped
+	tq.stopped = true
+	return stopped
+}
+
+type mockAfter struct {
+	lock        *sync.Mutex
+	startTime   time.Time
+	currentTime time.Time
+	queue       []*timerQueueItem
+}
+
+func newMockAfter() *mockAfter {
+	return &mockAfter{
+		queue: make([]*timerQueueItem, 0),
+		lock:  &sync.Mutex{},
+	}
+}
+
+func (m *mockAfter) AfterFunc(d time.Duration, f func()) stoppable {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	item := &timerQueueItem{
+		f: f,
+		t: m.currentTime.Add(d),
+	}
+	m.queue = append(m.queue, item)
+	return item
+}
+
+func (m *mockAfter) warp(d time.Duration) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.currentTime = m.currentTime.Add(d)
+	for _, item := range m.queue {
+		if item.run || item.stopped {
+			continue
+		}
+
+		if item.t.Before(m.currentTime) {
+			item.run = true
+			go item.f()
+		}
+	}
+}
```

Comment:
So there's some pre-existing work in the `kubernetes/utils` project for fake clocks, which include functions like this (Warp) which allow you to manipulate time.

We are going to be using this to test aspects of the renewal/ACME validation controller, so it'd be brilliant if we can make a start here. I *thought* I had already begun this somewhere, but it must be lost in a branch... somewhere. I recall when I did do this, I hit an issue in the package relating to cancellation of timers, so I opened this PR to fix it: https://github.com/kubernetes/utils/pull/37

Could you take a look and see how viable it'd be to switch over to this package for clock manipulation?

I don't want to cause a load of extra work for you for a simple change, so we can revisit it if it'll be painful, but I definitely think we'd like to end out in a place where we use a single fake clock implementation for all tests that need it 😄

Reply:
The clock there doesn't have `AfterFunc` in its interface. It does have `After`, but the fake doesn't implement it.

I believe the code here would have to be refactored to use `NewTimer` directly, or the clock interface / fake would have to be updated to support AfterFunc.

Based on that, I think it would be non-trivial to use the fake clock here, so I'd rather make one of the above changes in the future.
I think switching away from `AfterFunc` is a reasonable thing to do, but I think it will require a more significant refactor (mainly to avoid leaking goroutines as is all to easy to do with NewTimer)